src/history/hs_rec.c

```diff
@@ -160,29 +158,25 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
     }
 
     /*
-     * If we're inserting a non-zero timestamp, look ahead for any higher timestamps. If we find
-     * updates, we should remove them and reinsert them at the current timestamp.
+     * Look ahead for any higher timestamps. If we find updates, we should remove them and reinsert
+     * them at the current timestamp. If there were no keys equal to or less than our target key, we
+     * would have received WT_NOT_FOUND. In that case we need to search again with a higher
+     * timestamp.
      */
-    if (tw->start_ts != WT_TS_NONE) {
-        /*
-         * If there were no keys equal to or less than our target key, we would have received
-         * WT_NOTFOUND. In that case we need to search again with a higher timestamp as the cursor
-         * would not be positioned correctly.
-         */
-        if (ret == 0)
-            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
-        else {
-            F_SET(cursor, WT_CURSTD_HS_READ_ALL);
+    if (ret == 0)
+        WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+    else {
+        F_SET(cursor, WT_CURSTD_HS_READ_ALL);
 
-            cursor->set_key(cursor, 3, btree->id, key, tw->start_ts + 1);
-            WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_after(session, cursor), true);
+        cursor->set_key(cursor, 3, btree->id, key, tw->start_ts + 1);
+        WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_after(session, cursor), true);
 
-            if (!hs_read_all_flag)
-                F_CLR(cursor, WT_CURSTD_HS_READ_ALL);
-        }
-        if (ret == 0)
-            WT_ERR(__hs_fixup_out_of_order_from_pos(
-              session, cursor, btree, key, tw->start_ts, &counter));
+        if (!hs_read_all_flag)
+            F_CLR(cursor, WT_CURSTD_HS_READ_ALL);
+    }
+    if (ret == 0) {
+        WT_ERR(__hs_delete_reinsert_from_pos(
+          session, cursor, btree->id, key, tw->start_ts + 1, true, &counter));
```

Comment:
We are setting `tw->start_ts + 1` here because we actually want to delete them from `tw->start_ts`?

Reply:
Yes, that's correct.